File: ios/NorthAx/Networking/NorthAxAPI.swift
import Foundation

/// Domain-typed facade over `APIClient`: every method returns the app's existing
/// models (mapped from DTOs), so the store and views speak domain types only.
struct NorthAxAPI {
    static let shared = NorthAxAPI()
    private let client = APIClient.shared

    // MARK: - Profile

    func updateProfileName(_ name: String) async throws {
        _ = try await client.send("PATCH", "user/profile", body: UpdateProfileRequest(name: name))
    }

    // MARK: - Readiness & metrics

    func readinessToday() async throws -> DailyReadiness {
        let dto: DailyReadinessResponse = try await client.get("readiness/today")
        return dto.toDomain()
    }

    func metricsToday() async throws -> TrainingMetrics {
        let dto: DailyMetricsResponse = try await client.get("metrics/daily")
        return dto.toDomain()
    }

    /// Submit user-entered wellness values as a `manual` source; the server
    /// re-resolves the day against all sources and returns the merged metrics.
    @discardableResult
    func submitManualMetrics(_ req: ManualMetricsRequest) async throws -> TrainingMetrics {
        let dto: DailyMetricsResponse = try await client.post("metrics/manual", body: req)
        return dto.toDomain()
    }

    // MARK: - Plan & preferences

    func plans(weeks: Int = 4) async throws -> [WeeklyPlan] {
        let dtos: [WeeklyPlanResponse] = try await client.get(
            "plan/weeks", query: [URLQueryItem(name: "weeks", value: String(weeks))]
        )
        return dtos.map { $0.toDomain() }
    }

    /// Generate the next two weeks with the AI planner. Falls back server-side to
    /// the deterministic engine, so this always returns a valid plan. Uses a long
    /// timeout — the model call can take a while.
    func generatePlanAI() async throws -> [WeeklyPlan] {
        let dtos: [WeeklyPlanResponse] = try await client.post("plan/generate-ai", timeout: 120)
        return dtos.map { $0.toDomain() }
    }

    /// Replace one day's session with a chosen alternative (§9). `weekStart` must
    /// be the Monday of the week containing `date`. Returns the updated week.
    func overrideDay(weekStart: Date, date: Date, suggestion: SwitchSuggestion) async throws -> WeeklyPlan {
        let session = PlannedSessionDTO(
            domain: suggestion.domain.rawValue,
            title: suggestion.title,
            subtitle: suggestion.description,
            duration: suggestion.duration,
            intensityLabel: suggestion.intensityLabel,
            workout: suggestion.workout,
            exercises: suggestion.exercises?.map {
                ExerciseDTO(name: $0.name, muscleGroup: $0.muscleGroup.rawValue,
                            sets: $0.sets, repsRange: $0.repsRange, rest: $0.rest, notes: $0.notes)
            }
        )
        let ds = JSONCoders.calendarDate
        let dto: WeeklyPlanResponse = try await client.patch(
            "plan/week/\(ds.string(from: weekStart))/day/\(ds.string(from: date))",
            body: DayOverrideRequest(session: session)
        )
        return dto.toDomain()
    }

    func preferences() async throws -> ParsedPreferences {
        let dto: UserPreferencesDTO = try await client.get("preferences")
        return dto.toDomain()
    }

    /// Replace the per-sport weekly schedule; the server regenerates forward weeks.
    func updateSchedule(_ schedules: [DomainSchedule]) async throws -> ParsedPreferences {
        let dtos = schedules.map { DomainScheduleDTO(domain: $0.domain.rawValue, weekdays: $0.weekdays.sorted()) }
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/schedule", body: SchedulePatch(domainSchedules: dtos)
        )
        return dto.toDomain()
    }

    /// Partial merge of athlete thresholds; does NOT regenerate plans.
    func updateThresholds(_ partial: AthleteThresholds) async throws -> ParsedPreferences {
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/thresholds", body: ThresholdsPatch(thresholds: partial.toDTO())
        )
        return dto.toDomain()
    }

    func updateMuscleSplit(_ split: WeeklyMuscleGroupSplit) async throws -> ParsedPreferences {
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/muscle-split", body: MuscleSplitPatch(muscleGroupSplit: split.toDTO())
        )
        return dto.toDomain()
    }

    func updateDomains(_ domains: [TrainingDomain]) async throws -> ParsedPreferences {
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/domains", body: DomainsPatch(enabledDomains: domains.map(\.rawValue))
        )
        return dto.toDomain()
    }

    /// Set the cycling structured-workout target ("hr" | "power"); regenerates plans.
    func updateCyclingTarget(_ target: String) async throws -> ParsedPreferences {
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/target", body: CyclingTargetPatch(cyclingTarget: target)
        )
        return dto.toDomain()
    }

    /// Replace the per-sport goal targets (staged plan change; no server-side regen —
    /// the caller follows up with `generatePlanAI()`).
    func updateSportTargets(_ targets: [TrainingDomain: SportTarget]) async throws -> ParsedPreferences {
        let wire = Dictionary(uniqueKeysWithValues: targets.map { ($0.key.rawValue, $0.value.toDTO()) })
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/sport-targets", body: SportTargetsPatch(sportTargets: wire)
        )
        return dto.toDomain()
    }

    /// Latest AI goal-progress verdict per targeted sport (empty when none).
    func goalProgress() async throws -> [GoalCheck] {
        let dtos: [GoalProgressDTO] = try await client.get("goals/progress")
        return dtos.compactMap { $0.toDomain() }
    }

    /// Sync the per-metric source ranking used for multi-integration conflict
    /// resolution; does not regenerate plans.
    func updateMetricPriority(_ priority: MetricSourcePriority) async throws -> ParsedPreferences {
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/metric-priority", body: MetricPriorityPatch(metricPriority: priority.wire)
        )
        return dto.toDomain()
    }

    /// Sync the activity-source preference used to de-duplicate cross-source
    /// workouts (§13); does not regenerate plans.
    func updateActivityPriority(_ priority: ActivitySourcePriority) async throws -> ParsedPreferences {
        let dto: UserPreferencesDTO = try await client.patch(
            "preferences/activity-priority", body: ActivityPriorityPatch(activityPriority: priority.wire)
        )
        return dto.toDomain()
    }

    // MARK: - Coach

    func coachHistory(limit: Int = 50) async throws -> [CoachMessage] {
        let dtos: [CoachMessageDTO] = try await client.get(
            "ai/coach/history", query: [URLQueryItem(name: "limit", value: String(limit))]
        )
        return dtos.map { $0.toDomain() }
    }

    func clearCoachHistory() async throws {
        _ = try await client.send("DELETE", "ai/coach/history")
    }

    // MARK: - Strength

    /// Pre-fetched AI alternatives for one planned session (§9). Returns [] on
    /// failure so the caller can fall back to the deterministic switcher.
    func switchSuggestions(session: PlannedSession, date: Date) async throws -> [SwitchSuggestion] {
        let body = SwitchSuggestionRequest(
            domain: session.domain.rawValue, title: session.title,
            duration: session.duration, intensityLabel: session.intensityLabel,
            date: JSONCoders.calendarDate.string(from: date)
        )
        let resp: SwitchSuggestionsResponse = try await client.post(
            "ai/switch-suggestions", body: body, timeout: 120
        )
        return resp.suggestions.compactMap { $0.toDomain() }
    }

    func strengthSession(muscleGroups: [MuscleGroup], readinessScore: Int?) async throws -> StrengthSession {
        let dto: StrengthSessionResponse = try await client.post(
            "ai/strength/generate",
            body: StrengthGenerateRequest(
                muscleGroups: muscleGroups.map(\.rawValue), readinessScore: readinessScore
            )
        )
        return dto.toDomain()
    }

    // MARK: - Garmin

    func intervalsStatus() async throws -> IntervalsConnectionState {
        let dto: IntervalsStatusDTO = try await client.get("intervals/status")
        return dto.toConnectionState()
    }

    func intervalsAuthorizationURL() async throws -> URL {
        let dto: IntervalsConnectResponse = try await client.post("intervals/connect")
        guard let url = URL(string: dto.authorizationUrl) else { throw APIError.decoding }
        return url
    }

    /// Connect using a personal intervals.icu API key (HTTP Basic on the server).
    func connectWithAPIKey(athleteId: String, apiKey: String) async throws -> IntervalsConnectionState {
        let dto: IntervalsStatusDTO = try await client.post(
            "intervals/connect/apikey",
            body: IntervalsApiKeyConnect(athleteId: athleteId, apiKey: apiKey)
        )
        return dto.toConnectionState()
    }

    @discardableResult
    func intervalsSync() async throws -> Bool {
        _ = try await client.send("POST", "intervals/sync")
        return true
    }

    func intervalsDisconnect() async throws {
        _ = try await client.send("DELETE", "intervals/disconnect")
    }

    /// Time-series streams for a completed activity (§10). Empty arrays when the
    /// integration has no streams for it.
    func activityStreams(activityId: String) async throws -> ActivityStreams {
        // Source-aware: the backend resolves the activity's source (intervals/Strava).
        let dto: ActivityStreamsDTO = try await client.get("activities/\(activityId)/streams")
        return dto.toDomain()
    }

    /// Strava segment efforts for a completed activity (§13). Empty when none.
    func activitySegments(activityId: String) async throws -> [SegmentEffort] {
        let dtos: [SegmentEffortDTO] = try await client.get("activities/\(activityId)/segments")
        return dtos.map { $0.toDomain() }
    }

    /// The athlete's effort history on one segment, newest first.
    func segmentHistory(segmentId: String) async throws -> SegmentHistory {
        let dto: SegmentHistoryDTO = try await client.get("segments/\(segmentId)/efforts")
        return dto.toDomain()
    }

    // MARK: - Strava (§13)

    func stravaStatus() async throws -> IntervalsConnectionState {
        let dto: IntervalsStatusDTO = try await client.get("integrations/strava/status")
        return dto.toConnectionState()
    }

    /// Connect the single athlete via the server's personal refresh token (no redirect).
    func stravaConnectPersonal() async throws -> IntervalsConnectionState {
        let dto: IntervalsStatusDTO = try await client.post("integrations/strava/connect/personal", timeout: 30)
        return dto.toConnectionState()
    }

    @discardableResult
    func stravaSync() async throws -> Bool {
        _ = try await client.send("POST", "integrations/strava/sync")
        return true
    }

    func stravaDisconnect() async throws {
        _ = try await client.send("DELETE", "integrations/strava/disconnect")
    }

    /// One bounded batch of segment-effort backfill for existing Strava
    /// activities; repeat until `remaining == 0`.
    func stravaSegmentsBackfill() async throws -> StravaSegmentsBackfillDTO {
        try await client.post("integrations/strava/segments/backfill", timeout: 60)
    }

    /// Persist an in-app logged workout as a `manual` activity.
    @discardableResult
    func createActivity(_ request: ActivityCreateRequest) async throws -> GarminActivity {
        let dto: ActivityDTO = try await client.post("activities", body: request)
        return dto.toGarminActivity()
    }

    /// Rewrite the exercise log of an in-app logged (manual) strength activity.
    @discardableResult
    func updateActivityExercises(id: String, exercises: [LoggedExerciseDTO]) async throws -> GarminActivity {
        let dto: ActivityDTO = try await client.patch(
            "activities/\(id)", body: ActivityExercisesPatch(strengthExercises: exercises)
        )
        return dto.toGarminActivity()
    }

    func activities(limit: Int = 20, source: String? = nil) async throws -> [GarminActivity] {
        var query = [URLQueryItem(name: "limit", value: String(limit))]
        if let source { query.append(URLQueryItem(name: "source", value: source)) }
        let page: PaginatedActivities = try await client.get("activities", query: query)
        return page.items.map { $0.toGarminActivity() }
    }

    @discardableResult
    func pushWorkout(date: Date, session: PlannedSession) async throws -> String {
        let payload = IntervalsWorkoutPushRequest(
            date: JSONCoders.calendarDate.string(from: date),
            session: PlannedSessionDTO(
                domain: session.domain.rawValue, title: session.title,
                subtitle: session.subtitle, duration: session.duration,
                intensityLabel: session.intensityLabel, workout: session.workout
            )
        )
        let resp: IntervalsWorkoutPushResponse = try await client.post("intervals/workouts/push", body: payload)
        return resp.workoutId
    }
}

File: android/app/src/main/java/app/northax/data/remote/NorthAxApi.kt
package app.northax.data.remote

import app.northax.data.ParsedPreferences
import app.northax.data.remote.dto.ActivityCreateRequest
import app.northax.data.remote.dto.ActivityDto
import app.northax.data.remote.dto.ActivityExercisesPatch
import app.northax.data.remote.dto.ActivityPriorityPatch
import app.northax.data.remote.dto.LoggedExerciseDto
import app.northax.data.remote.dto.ActivityStreamsDto
import app.northax.data.remote.dto.CyclingTargetPatch
import app.northax.data.remote.dto.DailyMetricsResponse
import app.northax.data.remote.dto.DailyReadinessResponse
import app.northax.data.remote.dto.DayOverrideRequest
import app.northax.data.remote.dto.DomainScheduleDto
import app.northax.data.remote.dto.DomainsPatch
import app.northax.data.remote.dto.ExerciseDto
import app.northax.data.remote.dto.GoalProgressDto
import app.northax.data.remote.dto.IntervalsApiKeyConnect
import app.northax.data.remote.dto.IntervalsConnectResponse
import app.northax.data.remote.dto.IntervalsStatusDto
import app.northax.data.remote.dto.IntervalsWorkoutPushRequest
import app.northax.data.remote.dto.ManualMetricsRequest
import app.northax.data.remote.dto.MetricPriorityPatch
import app.northax.data.remote.dto.MuscleSplitPatch
import app.northax.data.remote.dto.CoachMessageDto
import app.northax.data.remote.dto.PaginatedActivities
import app.northax.data.remote.dto.PlannedSessionDto
import app.northax.data.remote.dto.SchedulePatch
import app.northax.data.remote.dto.SportTargetsPatch
import app.northax.data.remote.dto.StrengthGenerateRequest
import app.northax.data.remote.dto.StrengthSessionResponse
import app.northax.data.remote.dto.SwitchSuggestionRequest
import app.northax.data.remote.dto.SwitchSuggestionsResponse
import app.northax.data.remote.dto.ThresholdsPatch
import app.northax.data.remote.dto.UpdateProfileRequest
import app.northax.data.remote.dto.UserPreferencesDto
import app.northax.data.remote.dto.WeeklyPlanResponse
import app.northax.data.toConnectionState
import app.northax.data.toDomain
import app.northax.data.toDto
import app.northax.data.toGarminActivity
import app.northax.domain.model.ActivitySourcePriority
import app.northax.domain.model.ActivityStreams
import app.northax.domain.model.AthleteThresholds
import app.northax.domain.model.CoachMessage
import app.northax.domain.model.DailyReadiness
import app.northax.domain.model.DomainSchedule
import app.northax.domain.model.GarminActivity
import app.northax.domain.model.GoalCheck
import app.northax.domain.model.IntervalsConnectionState
import app.northax.domain.model.MetricSourcePriority
import app.northax.domain.model.MuscleGroup
import app.northax.domain.model.PlannedSession
import app.northax.domain.model.SportTarget
import app.northax.domain.model.StrengthSession
import app.northax.domain.model.SwitchSuggestion
import app.northax.domain.model.TrainingDomain
import app.northax.domain.model.TrainingMetrics
import app.northax.domain.model.WeeklyMuscleGroupSplit
import app.northax.domain.model.WeeklyPlan
import java.time.LocalDate

/**
 * Domain-typed facade over [ApiClient]: every method returns the app's domain
 * models (mapped from DTOs), so the store and views speak domain types only —
 * a 1:1 port of the iOS NorthAxAPI.
 */
class NorthAxApi(private val client: ApiClient) {

    // MARK: - Profile

    suspend fun updateProfileName(name: String) {
        client.send("PATCH", "user/profile", UpdateProfileRequest(name))
    }

    // MARK: - Readiness & metrics

    suspend fun readinessToday(): DailyReadiness =
        client.get<DailyReadinessResponse>("readiness/today").toDomain()

    suspend fun metricsToday(): TrainingMetrics =
        client.get<DailyMetricsResponse>("metrics/daily").toDomain()

    /** Submit user-entered wellness values as a `manual` source; the server
     *  re-resolves the day against all sources and returns the merged metrics. */
    suspend fun submitManualMetrics(req: ManualMetricsRequest): TrainingMetrics =
        client.post<DailyMetricsResponse, ManualMetricsRequest>("metrics/manual", req).toDomain()

    // MARK: - Plan & preferences

    suspend fun plans(weeks: Int = 4): List<WeeklyPlan> =
        client.get<List<WeeklyPlanResponse>>("plan/weeks", mapOf("weeks" to weeks.toString()))
            .map { it.toDomain() }

    /** Generate the next two weeks with the AI planner. Falls back server-side
     *  to the deterministic engine. Uses a long timeout. */
    suspend fun generatePlanAI(): List<WeeklyPlan> =
        client.post<List<WeeklyPlanResponse>, Unit>("plan/generate-ai", null, timeoutSeconds = 120)
            .map { it.toDomain() }

    /** Replace one day's session with a chosen alternative. `weekStart` must
     *  be the Monday of the week containing `date`. Returns the updated week. */
    suspend fun overrideDay(weekStart: LocalDate, date: LocalDate, suggestion: SwitchSuggestion): WeeklyPlan {
        val session = PlannedSessionDto(
            domain = suggestion.domain.raw,
            title = suggestion.title,
            subtitle = suggestion.description,
            duration = suggestion.duration,
            intensityLabel = suggestion.intensityLabel,
            workout = suggestion.workout,
            exercises = suggestion.exercises?.map {
                ExerciseDto(it.name, it.muscleGroup.raw, it.sets, it.repsRange, it.rest, it.notes)
            },
        )
        return client.patch<WeeklyPlanResponse, DayOverrideRequest>(
            "plan/week/$weekStart/day/$date",
            DayOverrideRequest(session),
        ).toDomain()
    }

    suspend fun preferences(): ParsedPreferences =
        client.get<UserPreferencesDto>("preferences").toDomain()

    /** Replace the per-sport weekly schedule; the server regenerates forward weeks. */
    suspend fun updateSchedule(schedules: List<DomainSchedule>): ParsedPreferences =
        client.patch<UserPreferencesDto, SchedulePatch>(
            "preferences/schedule",
            SchedulePatch(schedules.map { DomainScheduleDto(it.domain.raw, it.weekdays.sorted()) }),
        ).toDomain()

    /** Partial merge of athlete thresholds; does NOT regenerate plans. */
    suspend fun updateThresholds(partial: AthleteThresholds): ParsedPreferences =
        client.patch<UserPreferencesDto, ThresholdsPatch>(
            "preferences/thresholds", ThresholdsPatch(partial.toDto()),
        ).toDomain()

    suspend fun updateMuscleSplit(split: WeeklyMuscleGroupSplit): ParsedPreferences =
        client.patch<UserPreferencesDto, MuscleSplitPatch>(
            "preferences/muscle-split", MuscleSplitPatch(split.toDto()),
        ).toDomain()

    suspend fun updateDomains(domains: List<TrainingDomain>): ParsedPreferences =
        client.patch<UserPreferencesDto, DomainsPatch>(
            "preferences/domains", DomainsPatch(domains.map { it.raw }),
        ).toDomain()

    /** Set the cycling structured-workout target ("hr" | "power"); regenerates plans. */
    suspend fun updateCyclingTarget(target: String): ParsedPreferences =
        client.patch<UserPreferencesDto, CyclingTargetPatch>(
            "preferences/target", CyclingTargetPatch(target),
        ).toDomain()

    /** Replace the per-sport goal targets (staged plan change; the caller
     *  follows up with [generatePlanAI]). */
    suspend fun updateSportTargets(targets: Map<TrainingDomain, SportTarget>): ParsedPreferences =
        client.patch<UserPreferencesDto, SportTargetsPatch>(
            "preferences/sport-targets",
            SportTargetsPatch(targets.entries.associate { (k, v) -> k.raw to v.toDto() }),
        ).toDomain()

    /** Latest AI goal-progress verdict per targeted sport (empty when none). */
    suspend fun goalProgress(): List<GoalCheck> =
        client.get<List<GoalProgressDto>>("goals/progress").mapNotNull { it.toDomain() }

    /** Sync the per-metric source ranking; does not regenerate plans. */
    suspend fun updateMetricPriority(priority: MetricSourcePriority): ParsedPreferences =
        client.patch<UserPreferencesDto, MetricPriorityPatch>(
            "preferences/metric-priority", MetricPriorityPatch(priority.wire),
        ).toDomain()

    /** Sync the activity-source preference used to de-duplicate cross-source
     *  workouts; does not regenerate plans. */
    suspend fun updateActivityPriority(priority: ActivitySourcePriority): ParsedPreferences =
        client.patch<UserPreferencesDto, ActivityPriorityPatch>(
            "preferences/activity-priority", ActivityPriorityPatch(priority.wire),
        ).toDomain()

    // MARK: - Coach

    suspend fun coachHistory(limit: Int = 50): List<CoachMessage> =
        client.get<List<CoachMessageDto>>("ai/coach/history", mapOf("limit" to limit.toString()))
            .map { it.toDomain() }

    suspend fun clearCoachHistory() {
        client.send("DELETE", "ai/coach/history")
    }

    // MARK: - Strength & switches

    /** Pre-fetched AI alternatives for one planned session. */
    suspend fun switchSuggestions(session: PlannedSession, date: LocalDate): List<SwitchSuggestion> {
        val body = SwitchSuggestionRequest(
            domain = session.domain.raw, title = session.title,
            duration = session.duration, intensityLabel = session.intensityLabel,
            date = date.toString(),
        )
        return client.post<SwitchSuggestionsResponse, SwitchSuggestionRequest>(
            "ai/switch-suggestions", body, timeoutSeconds = 120,
        ).suggestions.mapNotNull { it.toDomain() }
    }

    suspend fun strengthSession(muscleGroups: List<MuscleGroup>, readinessScore: Int?): StrengthSession =
        client.post<StrengthSessionResponse, StrengthGenerateRequest>(
            "ai/strength/generate",
            StrengthGenerateRequest(muscleGroups.map { it.raw }, readinessScore),
        ).toDomain()

    // MARK: - intervals.icu

    suspend fun intervalsStatus(): IntervalsConnectionState =
        client.get<IntervalsStatusDto>("intervals/status").toConnectionState()

    suspend fun intervalsAuthorizationUrl(): String =
        client.post<IntervalsConnectResponse, Unit>("intervals/connect", null).authorizationUrl

    /** Connect using a personal intervals.icu API key (HTTP Basic on the server). */
    suspend fun connectWithApiKey(athleteId: String, apiKey: String): IntervalsConnectionState =
        client.post<IntervalsStatusDto, IntervalsApiKeyConnect>(
            "intervals/connect/apikey", IntervalsApiKeyConnect(athleteId, apiKey),
        ).toConnectionState()

    suspend fun intervalsSync() {
        client.send("POST", "intervals/sync")
    }

    suspend fun intervalsDisconnect() {
        client.send("DELETE", "intervals/disconnect")
    }

    /** Time-series streams for a completed activity. Empty arrays when the
     *  integration has no streams for it. */
    suspend fun activityStreams(activityId: String): ActivityStreams =
        client.get<ActivityStreamsDto>("activities/$activityId/streams").toDomain()

    // MARK: - Strava

    suspend fun stravaStatus(): IntervalsConnectionState =
        client.get<IntervalsStatusDto>("integrations/strava/status").toConnectionState()

    /** Connect the single athlete via the server's personal refresh token (no redirect). */
    suspend fun stravaConnectPersonal(): IntervalsConnectionState =
        client.post<IntervalsStatusDto, Unit>("integrations/strava/connect/personal", null, timeoutSeconds = 30)
            .toConnectionState()

    suspend fun stravaSync() {
        client.send("POST", "integrations/strava/sync")
    }

    suspend fun stravaDisconnect() {
        client.send("DELETE", "integrations/strava/disconnect")
    }

    // MARK: - Activities

    /** Persist an in-app logged workout as a `manual` activity. */
    suspend fun createActivity(request: ActivityCreateRequest): GarminActivity =
        client.post<ActivityDto, ActivityCreateRequest>("activities", request).toGarminActivity()

    /** Rewrite the exercise log of an in-app logged (manual) strength activity. */
    suspend fun updateActivityExercises(id: String, exercises: List<LoggedExerciseDto>): GarminActivity =
        client.patch<ActivityDto, ActivityExercisesPatch>("activities/$id", ActivityExercisesPatch(exercises))
            .toGarminActivity()

    suspend fun activities(limit: Int = 20, source: String? = null): List<GarminActivity> {
        val query = buildMap {
            put("limit", limit.toString())
            source?.let { put("source", it) }
        }
        return client.get<PaginatedActivities>("activities", query).items.map { it.toGarminActivity() }
    }

    suspend fun pushWorkout(date: LocalDate, session: PlannedSession): String {
        val payload = IntervalsWorkoutPushRequest(
            date = date.toString(),
            session = PlannedSessionDto(
                domain = session.domain.raw, title = session.title,
                subtitle = session.subtitle, duration = session.duration,
                intensityLabel = session.intensityLabel, workout = session.workout,
            ),
        )
        return client.post<app.northax.data.remote.dto.IntervalsWorkoutPushResponse, IntervalsWorkoutPushRequest>(
            "intervals/workouts/push", payload,
        ).workoutId
    }
}
